File: 0496-Next Greater Element I/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std; 

// nums1只是拿來索引 
// 另一種想法 先處理nums2, 然後再利用處理完的vector根據num1來解出solution
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> hash;
        
        for(int idx = 0; idx < nums1.size(); ++idx) {
            hash[nums1[idx]] = idx; // make val->index hash;
        }
        
        stack<int> s;
        vector<int> solution(nums1.size(), -1);
        
        for(int idx = 0; idx < nums2.size(); ++idx) {
            int val = nums2[idx];
            while(!s.empty() && val > nums2[s.top()]) {
                int index = s.top();
                if(hash.find(nums2[index]) != hash.end()) {
                    solution[hash[nums2[index]]] = val;
                }
                s.pop();
            }
            s.push(idx);
        }
        
        
        return solution;
    }
};

// Runtime: 8 ms, faster than 97.15% of C++ online submissions for Next Greater Element I.
// Memory Usage: 9.1 MB, less than 65.70% of C++ online submissions for Next Greater Element I.

File: 0496-Next Greater Element I/golang/solution.go
package main

import "fmt"

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	var stack []int
	tmp := make(map[int]int)

	for idx, val := range nums2 {
		// while(!stack.empty() && val > stack.top() element)
		for len(stack) != 0 && val > nums2[stack[len(stack)-1]] {
			topIndex := len(stack) - 1
			tmp[nums2[stack[topIndex]]] = val
			stack = stack[:len(stack)-1] // pop
		}
		stack = append(stack, idx) // push
	}

	solution := make([]int, len(nums1))

	for idx, val := range nums1 {
		if v, ok := tmp[val]; ok {
			solution[idx] = v
		} else {
			solution[idx] = -1
		}
	}

	return solution
}

func main() {
	s := nextGreaterElement([]int{4, 1, 2}, []int{1, 3, 4, 2})
	fmt.Println(s)
}
